things/postgres/groups.go

```diff
@@ -0,0 +1,636 @@
+	}
+
+	rows, err := gr.db.NamedQueryContext(ctx, q, dbPage)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+	defer rows.Close()
+
+	items, err := processRows(rows)
+	if err != nil {
+		return groups.GroupPage{}, err
+	}
+
+	total, err := total(ctx, gr.db, cq, dbPage)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+
+	page := groups.GroupPage{
+		Groups: items,
+		PageMetadata: groups.PageMetadata{
+			Total: total,
+		},
+	}
+
+	return page, nil
+}
+
+func (gr groupRepository) RetrieveAllParents(ctx context.Context, groupID string, level uint64, gm groups.Metadata) (groups.GroupPage, error) {
+	if groupID == "" {
+		return groups.GroupPage{}, nil
+	}
+
+	_, mq, err := getGroupsMetadataQuery("thing_groups", gm)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errRetrieveDB, err)
+	}
+	if mq != "" {
+		mq = fmt.Sprintf("AND %s", mq)
+	}
+
+	q := fmt.Sprintf(`SELECT id, name, owner_id, parent_id, description, metadata, path, nlevel(path) as level FROM thing_groups 
+					  WHERE path @> :path AND nlevel(:path) - nlevel(path) >= 0 AND nlevel(:path) - nlevel(path) <= :level %s`, mq)
+	cq := fmt.Sprintf("SELECT COUNT(*) FROM thing_groups WHERE path @> :path AND nlevel(:path) - nlevel(path) >= 0 AND nlevel(:path) - nlevel(path) <= :level %s", mq)
+	g, err := gr.RetrieveByID(ctx, groupID)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(groups.ErrNotFound, err)
+	}
+	if level > maxLevel {
+		level = maxLevel
+	}
+
+	dbPage, err := toDBGroupPage("", g.Path, level, gm)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+
+	rows, err := gr.db.NamedQueryContext(ctx, q, dbPage)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+	defer rows.Close()
+
+	items, err := processRows(rows)
+	if err != nil {
+		return groups.GroupPage{}, err
+	}
+
+	total, err := total(ctx, gr.db, cq, dbPage)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+
+	page := groups.GroupPage{
+		Groups: items,
+		PageMetadata: groups.PageMetadata{
+			Total: total,
+		},
+	}
+
+	return page, nil
+}
+
+func (gr groupRepository) RetrieveAllChildren(ctx context.Context, groupID string, level uint64, gm groups.Metadata) (groups.GroupPage, error) {
+	if groupID == "" {
+		return groups.GroupPage{}, nil
+	}
+	_, mq, err := getGroupsMetadataQuery("thing_groups", gm)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errRetrieveDB, err)
+	}
+	if mq != "" {
+		mq = fmt.Sprintf("AND %s", mq)
+	}
+
+	q := fmt.Sprintf(`SELECT id, name, owner_id, parent_id, description, metadata, path, nlevel(path) as level FROM thing_groups 
+					  WHERE path <@ :path AND nlevel(path) - nlevel(:path) <= :level %s`, mq)
+	cq := fmt.Sprintf("SELECT COUNT(*) FROM thing_groups WHERE path <@ :path AND nlevel(path) - nlevel(:path) <= :level %s", mq)
+	g, err := gr.RetrieveByID(ctx, groupID)
```

Comment:
Looks like you are fetching the group only to fetch a path for the response. Can this call be avoided (it's a separate transaction - one more DB driver call; also, because not everything is done in the transactions and can change parent in the current implementation - I'm afraid that we need an optimistic lock here)?